driver/src/driver/stablex_driver.rs

```diff
@@ -14,35 +15,61 @@ pub struct StableXDriver<'a> {
     past_auctions: HashSet<U256>,
     contract: &'a dyn StableXContract,
     price_finder: &'a mut dyn PriceFinding,
+    metrics: StableXMetrics,
 }
 
 impl<'a> StableXDriver<'a> {
-    pub fn new(contract: &'a dyn StableXContract, price_finder: &'a mut dyn PriceFinding) -> Self {
+    pub fn new(
+        contract: &'a dyn StableXContract,
+        price_finder: &'a mut dyn PriceFinding,
+        metrics: StableXMetrics,
+    ) -> Self {
         StableXDriver {
             past_auctions: HashSet::new(),
             contract,
             price_finder,
+            metrics,
         }
     }
 
     pub fn run(&mut self) -> Result<bool, DriverError> {
         // Try to process previous batch auction
-        let batch = self.contract.get_current_auction_index()? - 1;
-        if self.past_auctions.contains(&batch) {
+        let batch_to_solve_result = self
+            .contract
+            .get_current_auction_index()
+            .map(|batch_collecting_orders| batch_collecting_orders - 1);
+
+        if batch_to_solve_result
+            .as_ref()
+            .map(|batch| self.past_auctions.contains(batch))
+            .unwrap_or(false)
+        {
             return Ok(false);
         }
-        let (account_state, orders) = self.contract.get_auction_data(batch)?;
+
+        self.metrics
+            .auction_processing_started(&batch_to_solve_result);
+        let batch_to_solve = batch_to_solve_result?;
+
+        let get_auction_data_result = self.contract.get_auction_data(batch_to_solve);
+        self.metrics
+            .auction_orders_fetched(batch_to_solve, &get_auction_data_result);
+        let (account_state, orders) = get_auction_data_result?;
+
         let solution = if orders.is_empty() {
-            info!("No orders in batch {}", batch);
+            info!("No orders in batch {}", batch_to_solve);
             Solution::trivial(0)
         } else {
-            match self.price_finder.find_prices(&orders, &account_state) {
+            let price_finder_result = self.price_finder.find_prices(&orders, &account_state);
+            self.metrics
+                .auction_solution_computed(batch_to_solve, &orders, &price_finder_result);
+            match price_finder_result {
                 Ok(solution) => {
                     info!("Computed solution: {:?}", &solution);
                     solution
                 }
                 Err(err) => {
-                    self.past_auctions.insert(batch);
+                    self.past_auctions.insert(batch_to_solve);
```

Comment:
Somewhat unrelated; Is this "past_auctions" something that is stored in memory? If so, what happens if our services go down between batches? Will it try to solve an already finished batch and fail on restart?

Reply:
It will try to resolve it but likely not come up with a better solution and thus not resubmit the solution.